src/tool_operate.c

```diff
@@ -656,16 +656,106 @@ static void single_transfer_cleanup(struct OperationConfig *config)
   }
 }
 
+static char *ipfs_gateway(void)
+{
+  char *gateway;
+  char *ipfs_path;
+  char *gateway_file_path;
+  FILE *gateway_file;
+
+  gateway = curlx_getenv("IPFS_GATEWAY");
+
+  /* Gateway is found from environment variable. */
+  if(gateway) {
+    bool add_slash = (gateway[strlen(gateway) - 1] == '/') ? FALSE : TRUE;
+    gateway = aprintf("%s%s", gateway, (add_slash) ? "/" : "");
+    return gateway;
+  }
+
+  /* Try to find the gateway in the IPFS data folder. */
+
+  ipfs_path = curlx_getenv("IPFS_PATH");
+
+  if(!ipfs_path) {
+    /* Empty path, fallback to "~/.ipfs", as that's the default location. */
+    ipfs_path = aprintf("%s/.ipfs/", curlx_getenv("HOME"));
+  }
+
+  gateway_file_path = aprintf("%sgateway", ipfs_path);
+  gateway_file = fopen(gateway_file_path, FOPEN_READTEXT);
+
+  if(gateway_file) {
+    char gateway_buffer[255];
+    if(!fgets(gateway_buffer, sizeof(gateway_buffer) - 1, gateway_file)) {
+      fclose(gateway_file);
+      return NULL;
+    }
+    fclose(gateway_file);
+
+    /* Replace the last char with \0 */
+    gateway_buffer[sizeof(gateway_buffer) - 1] = 0;
+
+    /* Replace first occurence of end of line with \0 */
+    gateway_buffer[strcspn(gateway_buffer, "\r")] = 0;
+    gateway_buffer[strcspn(gateway_buffer, "\n")] = 0;
```

Comment:
This seems very strict. What if the file has no CR and only LF or neither? What if there's a whitespace? Where's the spec for this new file format you add support for here?